### PR TITLE
QtWebEngine

### PR DESCRIPTION
**Pull request by u-235**
Hey!

This patch contains quite a few changes that are closely related to each other and therefore it is difficult to separate them into several parts.

- I used GitHub Actions to test the changes on various configurations.
- The git submodule for lib/CHMLib made it possible to simplify GitHub Actions and builds on Windows. At the same time, the possibility of dynamic linking with libchm has been preserved.
- The source code for QtWebKit has received minor changes related to moving files to the webkit folder.
- For QtWebEngine, the sources were moved to the webengine folder and the viewwindowmgr.cpp and webenginepage.h files were added.
- Cmake now supports building with Qt4, Qt5 or KDE4 using QtWebKit or QtWebEngine. Qmake also supports building with both QtWebKit and QtWebEngine.

### CODE REVIEW

**Comment by gyunaev** [comment]
Unfortunately this cannot be merged as-is, as it includes many things which are completely independent. Please split it into different requests:

- Separate github actions/test request;
- Please make built-in CHMLib optional; on Linux most people would link with systemwide chmlib to keep them separate for security updates. It *must* be able to compile with system-wide chmlib.
- Please explain the purpose of all those changes in CMakefile as it is unclear what exactly you're trying to achieve. "Update CMake scripts" doesn't explain anything, I'm afraid. I.e. what's the purpose of changing SET to set in cmakefile?
- Please make the required change minimal. There are too many things which I'd consider style/formatting changes, which are unnecessary

**Review by gyunaev — COMMENTED** [review]
What is the purpose of this change? What problem does it solve?

File: lib/libebook/ebook_chm.cpp [hunk 1]
```diff
@@ -36,7 +36,7 @@
 //#define DEBUGPARSER(A)	qDebug A
 #define DEBUGPARSER(A)
 
-static const char * URL_SCHEME_CHM = "ms-its";
+const char * EBook_CHM::URL_SCHEME_CHM = "ms-its";
```
Comment (gyunaev) [inline lib/libebook/ebook_chm.cpp:39]: What issue are you solving here? Why is this change needed?

File: src/qtwebengine/dataprovider.cpp [hunk 2]
```diff
@@ -1,20 +1,66 @@
...
-#include "mainwindow.h"
-#include "dataprovider_qwebengine.h"
+#include "../mainwindow.h"
+#include "dataprovider.h"
+#include "ebook_chm.h"
+#include "ebook_epub.h"
+
+#define PRINT_DEBUG ( defined PRINT_DEBUG_ALL || defined PRINT_DEBUG_WEBENGINE || defined PRINT_DEBUG_WEBENGINEDATAPROVIDER )
+
+static struct RegistrationHelper
```
Comment (gyunaev) [inline src/qtwebengine/dataprovider.cpp:31]: This is definitely a class, not a structure

File: src/qtwebengine/dataprovider.cpp [hunk 3]
```diff
@@ -1,20 +1,66 @@
...
-    bool htmlfile = url.path().endsWith( ".html" ) || url.path().endsWith( ".htm" ) || url.path().endsWith( ".xhtml" );
-
-    qDebug("requestStarted %s", qPrintable(url.toString()) );
+    QByteArray headerAccept = request->requestHeaders().value( "Accept" );
+#if PRINT_DEBUG
+    qDebug() << "[DEBUG] DataProvider::requestStarted";
+    qDebug() << "  url = " << url.toString();
+    qDebug() << "  Header \"Accept\" = " << headerAccept;
+#endif
+    bool htmlfile = headerAccept.contains( "text/html" );
```
Comment (gyunaev) [inline src/qtwebengine/dataprovider.cpp:63]: Can you please point out to the documentation which states this works as expected? For me it did not work.

File: src/qtwebengine/dataprovider.cpp [hunk 4]
```diff
@@ -26,13 +72,23 @@ void DataProvider_QWebEngine::requestStarted( QWebEngineUrlRequestJob *request )
...
+    QByteArray mimetype;
 
-    // Specify the encoding in case the page content is not UTF-8
+    // We must specify the proper MIME type for the page to display correctly.
+    // The HTML and XML files correspond to "text/html";
+    // for other types "application/octet-stream" is sufficient.
+    // In addition, for "text/html", a "meta" tag is added specifying the text encoding.
+    // This is the easiest and most stable way to set the encoding.
     if ( htmlfile )
-        mimetype = QString( "text/html; charset=%1" ) .arg( ::mainWindow->chmFile()->currentEncoding() );
```
Comment (gyunaev) [inline src/qtwebengine/dataprovider.cpp]: Are you sure this works?

File: src/qtwebengine/viewwindow.cpp [hunk 5]
```diff
@@ -115,9 +108,7 @@ QMenu * ViewWindow::getContextMenu( const QUrl & link, QWidget * parent )
         {
             m_contextMenuLink = createStandardContextMenu( parent );
             m_contextMenuLink->addSeparator();
-
```
Comment (gyunaev) [inline src/qtwebengine/viewwindow.cpp:118]: Don't do this.

File: src/qtwebengine/dataprovider.h [hunk 6]
```diff
@@ -1,14 +1,36 @@
-#ifndef DATAPROVIDER_QWEBENGINE_H
-#define DATAPROVIDER_QWEBENGINE_H
+/*
+ *  Kchmviewer - a CHM and EPUB file viewer with broad language support
+ *  Copyright (C) 2004-2016 George Yunaev, gyunaev@ulduzsoft.com
+ *  Copyright (C) 2021 Nick Egorrov, nicegorov@yandex.ru
```
Comment (gyunaev) [inline src/qtwebengine/dataprovider.h:4]: This is rarely included in pull requests, honestly. Especially for minor changes.

File: src/qtwebengine/viewwindow.cpp [hunk 7]
```diff
@@ -246,9 +232,42 @@ void ViewWindow::updateHistoryIcons()
...
+\
+        el = el.parentElement;\
+    }\
+\
+    if (el.hasAttribute('href')){\
+        return el.href;\
+    } else {\
+        return null;\
+    }\
+}";
```
Comment (gyunaev) [inline src/qtwebengine/viewwindow.cpp:252]: A comment would be useful about what's the purpose of this code.

File: src/qtwebengine/viewwindow.cpp [hunk 8]
```diff
@@ -257,34 +276,37 @@ void ViewWindow::contextMenuEvent(QContextMenuEvent *e)
...
     delete m;
 }
 
 void ViewWindow::onLoadFinished ( bool )
 {
-/*    if ( m_storedScrollbarPosition > 0 )
+    if ( m_storedScrollbarPosition > 0 )
     {
-        page()->currentFrame()->setScrollBarValue( Qt::Vertical, m_storedScrollbarPosition );
+        page()->runJavaScript( QString( "document.body.scrollTop=%1" ).arg( m_storedScrollbarPosition )
```
Comment (gyunaev) [inline src/qtwebengine/viewwindow.cpp:289]: What gonna happen if pConfig->m_browserEnableJS is false (which is default)?

File: src/qtwebengine/dataprovider.cpp (hunk 3 above)
Comment (u-235) [reply src/qtwebengine/dataprovider.cpp:63]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept
https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types#texthtml

The [QWebEngineUrlRequestJob::requestHeaders()](https://doc.qt.io/qt-5/qwebengineurlrequestjob.html#requestHeaders) function was introduced in Qt 5.13, while I was aiming for the 5.9 minimal version. I guess I'll have to look for another solution.

Comment (gyunaev) [reply src/qtwebengine/dataprovider.cpp:63]: This is not what I asked. I asked if you're sure whether Accept: header is correctly set in this scenario in every single test case.

Comment (gyunaev) [reply src/qtwebengine/dataprovider.cpp:63]: I'm also puzzled, same as with many others, **why** this change was made? Lots of them seem to be made without any particular purpose. It is unclear what issue, if any, they fix.

File: src/qtwebengine/dataprovider.cpp [hunk 9]
```diff
@@ -26,13 +72,23 @@ void DataProvider_QWebEngine::requestStarted( QWebEngineUrlRequestJob *request )
...
+    {
+        mimetype = "text/html";
+        buf.prepend(QString( "<META http-equiv='Content-Type' content='text/html; charset=%1'>" )
+                    .arg( ::mainWindow->chmFile()->currentEncoding() ).toLatin1() );
```
Comment (gyunaev) [inline src/qtwebengine/dataprovider.cpp:86]: Note that your prepend will go before `<html>` tag, which may internally have another http-equiv.

File: src/qtwebengine/dataprovider.cpp [hunk 10]
```diff
@@ -41,8 +97,8 @@ void DataProvider_QWebEngine::requestStarted( QWebEngineUrlRequestJob *request )
     outbuf->close();
 
     // Only delete the buffer when the request is deleted too
-    connect( request, SIGNAL(destroyed()), outbuf, SLOT(deleteLater()) );
+    connect( request, SIGNAL( destroyed() ), outbuf, SLOT( deleteLater() ) );
```
Comment (gyunaev) [inline src/qtwebengine/dataprovider.cpp:100]: Please no. Any refactoring/reformatting MUST be a separate pull request, and should be discussed in advance if you want to have them integrated.

File: src/qtwebengine/dataprovider.cpp [hunk 11]
```diff
@@ -41,8 +97,8 @@ void DataProvider_QWebEngine::requestStarted( QWebEngineUrlRequestJob *request )
...
 
     // We're good to go
-    request->reply( "text/html", outbuf );
+    request->reply( mimetype, outbuf );
```
Comment (gyunaev) [inline src/qtwebengine/dataprovider.cpp:103]: This doesn't work, at least in my tests QtWebEngine simply ignores whatever you pass here.

File: src/qtwebengine/dataprovider.cpp (hunk 4 above)
Comment (u-235) [reply src/qtwebengine/dataprovider.cpp]: I tested the application in the following environment:

- CPU : Intel(R) Pentium(R) Dual CPU E2180 @ 2.00GHz
- RAM : 3GB
- OS : Linux Mint 18.1 Serena with the kernel "Linux 4.4.0-53-generic i686 GNU/Linux"
- DE : MATE Desktop Environment 1.16.2
- Qt : local build 5.15.2

And this case I checked especially carefully.

Comment (gyunaev) [reply src/qtwebengine/dataprovider.cpp]: Have you checked that the new code indeed changes anything within the detection?

File: src/qtwebengine/viewwindow.cpp (hunk 8 above)
Comment (u-235) [reply src/qtwebengine/viewwindow.cpp:289]: It's going to work.

Comment (gyunaev) [reply src/qtwebengine/viewwindow.cpp:289]: Have you tested it?
And, again, what issue are you fixing here? Are there situations when the above code isn't working?

File: src/qtwebengine/dataprovider.cpp (hunk 10 above)
Comment (u-235) [reply src/qtwebengine/dataprovider.cpp:100]: Sorry, it was an accident. Apparently, it happened until I disabled autoformatting in the IDE.

File: src/qtwebengine/viewwindow.cpp (hunk 8 above)
Comment (u-235) [reply src/qtwebengine/viewwindow.cpp:289]: No issues at all, except that QWebEnginePage::currentFrame() doesn't exist.

Comment (u-235) [reply src/qtwebengine/viewwindow.cpp:289]: Ладно, я понял, что мой код - унылое говно.

File: src/qtwebengine/dataprovider.cpp (hunk 11 above)
Comment (u-235) [reply src/qtwebengine/dataprovider.cpp:103]: Can I see your code with QtWebEngine?

File: src/qtwebengine/viewwindow.cpp (hunk 8 above)
Comment (gyunaev) [reply src/qtwebengine/viewwindow.cpp:289]: > No issues at all, except that QWebEnginePage::currentFrame() doesn't exist.

https://doc.qt.io/archives/qt-5.5/qwebpage.html#currentFrame

Comment (gyunaev) [reply src/qtwebengine/viewwindow.cpp:289]: > Ладно, я понял, что мой код - унылое говно.

Часть изменений выглядит неплохо, но их нужно отделить. Именно поэтому pull requests нужно делать небольшими, а не так что "вот все что я поменял, разбирайтесь сами, что тут нужно а что нет".

Comment (u-235) [reply src/qtwebengine/viewwindow.cpp:289]: > > No issues at all, except that QWebEnginePage::currentFrame() doesn't exist.
> 
> https://doc.qt.io/archives/qt-5.5/qwebpage.html#currentFrame

Yes, I read that documentation, but it doesn't apply to **QWebEnginePage**.

File: src/qtwebengine/viewwindow.cpp (hunk 5 above)
Comment (u-235) [reply src/qtwebengine/viewwindow.cpp:118]: Some files are indented with spaces, others with tabs, and a few have both spaces and tabs. There are also tail spaces. It's hard not to screw something up.

File: src/qtwebengine/viewwindow.cpp (hunk 8 above)
Comment (gyunaev) [reply src/qtwebengine/viewwindow.cpp:289]: Yep, the recommended way is indeed via JS: https://bugreports.qt.io/browse/QTBUG-48323

File: src/qtwebengine/viewwindow.cpp (hunk 5 above)
Comment (gyunaev) [reply src/qtwebengine/viewwindow.cpp:118]: This is why proofreading the .diff file is recommended. If this creates an issue, the proper solution is to submit a bug report, so this is fixed in the source tree.
